Modules/_threadmodule.c

```diff
@@ -1749,6 +1898,24 @@ thread_module_clear(PyObject *module)
     Py_CLEAR(state->lock_type);
     Py_CLEAR(state->local_type);
     Py_CLEAR(state->local_dummy_type);
+    // To avoid resource leaks, detach all still joinable threads
+    if (state->joinable_dict != NULL) {
+        PyObject *key, *value;
+        Py_ssize_t pos = 0;
+
+        while (PyDict_Next(state->joinable_dict, &pos, &key, &value)) {
+            if (PyLong_Check(value)) {
+                unsigned long long handle = PyLong_AsUnsignedLongLong(value);
+                if (handle == (unsigned long long) -1 && PyErr_Occurred()) {
+                    // Should not happen
+                    PyErr_Clear();
+                } else {
+                    PyThread_detach_thread((Py_uintptr_t) handle);
+                }
+            }
```

Comment:
We might also want to emit a ResourceWarning here, what do you think @vstinner ?

Reply:
Thinking about it, I'd rather defer this to a later PR (if at all).